src/persist/src/indexed/mod.rs

```diff
@@ -424,21 +410,103 @@ impl<K: Data, V: Data, U: Buffer, L: Blob> Indexed<K, V, U, L> {
         // moved is still there but now irrelevant. It may also be a good time
         // to compact trace.
+
+        // TODO: Instead of fully overwriting META each time, this should be
+        // more like a compactable log.
+        self.blob.set_meta(self.serialize_meta())?;
         Ok(())
     }
 
+    /// Returns the current "sealed" frontier for an id.
+    ///
+    /// This frontier represents a contract of time such that all updates with a
+    /// time less than it have arrived. This frontier is advanced though the
+    /// `seal` method. Once a time has been sealed for an id, it becomes an
+    /// error to later seal it at an time less than or equal to the sealed
+    /// frontier. It is also an error to write new data with a time less than
+    /// the sealed frontier.
+    fn sealed_frontier(&self, id: Id) -> Result<Antichain<u64>, Error> {
+        let trace = self
+            .traces
+            .get(&id)
+            .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
+        Ok(trace.seal())
+    }
+
+    /// Check if a given seal command is valid and if so, return the set of ids and
+    /// their previous seal frontiers.
+    fn validate_seal(&self, ids: &[Id], ts_upper: u64) -> Result<Vec<(Id, u64)>, Error> {
+        ids.iter()
+            .map(|id| {
+                let prev = self.sealed_frontier(*id)?;
+
+                if !prev.less_than(&ts_upper) {
+                    Err(Error::from(format!(
+                        "invalid seal for {:?}: {:?} not in advance of current seal frontier {:?}",
+                        id, ts_upper, prev
+                    )))
+                } else {
+                    Ok((*id, prev[0]))
+                }
+            })
+            .collect()
+    }
+
+    /// Sealing a time advances the "sealed" frontier for an id, which restricts
+    /// what times can later be sealed and written for that id. See
+    /// `sealed_frontier` for details.
+    pub fn seal(&mut self, ids: Vec<Id>, ts_upper: u64) -> Result<(), Error> {
+        let prev = self.validate_seal(&ids, ts_upper)?;
+
+        // Update in-memory state to reflect the new seal.
+        // TODO: we should keep the in-memory state immutable until after the
+        // results have been safely committed to durable storage. For now this
+        // protocol is simple enough to reason about.
+        for id in ids.iter() {
+            let trace = self.traces.get_mut(id).expect("trace known to exist");
+
+            trace.update_seal(ts_upper);
+        }
+
+        // TODO: Instead of fully overwriting META each time, this should be
+        // more like a compactable log.
+        match self.blob.set_meta(self.serialize_meta()) {
+            ok @ Ok(_) => {
+                // Notify listeners now that the seal has been committed to
+                // durable storage.
+                for id in ids.iter() {
+                    if let Some(listen_fns) = self.listeners.get(id) {
```

Comment:
I think we actually want to call this when we've successfully moved the data into trace. That way, a user that has received a seal knows that it can use any indexed access we add for trace data. The way this is now, some of that data might still be in future when the listener is notified

Reply:
I'm happy to make that change but then the behavior would be different between writes and seals because for writes we currently notify listeners as soon as the data lands in buffer. If its the same to you, I would vote that we leave it as is because the other pr which pushes all user requests into a command queue changes things quite a bit and makes it easier to notify listeners when updates get moved into trace.

Reply:
nvm! changed my mind and added it as a separate commit because its basically a prereq to having the other pr work correctly :D